### DISCUSSION TITLE
02/06일 멘토링 관련 질문 사항

### DISCUSSION THREAD

**Discussion by fkdl0048** [fkdl0048/Mentoring-Project#41] [Ideas]
공부하면서 생기는 궁금증들을 조금씩 적어놓겠습니다..!  

여유가 되신다면 2일전에 멘토링이 가능하신지 알려주시면 감사하겠습니다..!  

* 스스로 생각한 Todo스케줄 관련 질문(개발자로 공부해야하는 밸런스)  
지금 진행중인 공부 프로세스에 대한 조언을 받고 싶습니다.   

* 인터페이스 관련 질문
저번에 인터페이스를 조금 더 알아보는게 좋다고 하셔서 책에서 나온 내용 위주로 공부하고 있습니다.  
공부하면서 나온 궁금한 내용들 의견 나누고 싶습니다.

* 프로그래밍 패러다임 질문
프로그래밍 패러다임에서 헷갈리는 부분과 효과적인 학습방법  
다양한 프로그래밍 패러다임에 대한 생각과 유리한 지점의 의견을 나누고 싶습니다  

* 테스트코드란..?  
테스트코드의 중요성을 아직까지도 잘 모르는 것 같아서 조언을 구하고 싶습니다.  
책에서 말하는 이유와 방어적인 프로그래밍의 의도, 테스트 기반 프로그래밍도 궁금합니다.

* 진행중인 스터디 프로젝트 리뷰  
말씀해주신 개발 프로세스를 적극도입해보는 중인데 풀리퀘 단위에서 문제가 있습니다.

* [프로그래머의 길, 멘토에게 묻다] 의논
이 내용은 온라인 보다 오프라인으로 이야기 해보면 더 좋을 것 같습니다.  
다음주 모각코까지 진행하게 되면 다 읽게 되는데 이후 오프라인 멘토링 날짜가 잡히면 그때까지 질문 따로 정리해보겠습니다 

추가 예정

**Comment by jongfeel** [comment]
#### To do 스케줄

프로젝트는 1개 정도에 집중하고, 학교에 복학하면 학교 수업에 더 충실하는게 좋을 수 있습니다.
학기 중이 아니라면 2개 정도 까지는 가능할 수 있지만, 그래도 1개에 집중해서 성과를 내는 게 좋긴 합니다.

#### 인터페이스 관련

인터페이스를 사용하는데 이해하기 좋은 방법은 객체들 간의 의존성을 생각해 볼 때입니다.
Best practice로는 표준 라이브러리에서 인터페이스의 활용법을 검토해 보면 좋습니다.

#### 프로그래밍 패러다임

여러 가지가 있는데
크게는 명령형, 선언형으로 구분해 볼 수 있습니다.

명령형은 컴퓨터에게 구체 적인 코드를 작성하고 지시해서 수행하게 하는 방법입니다.
절차적이고 객체지향적인 방법에서 드러나는 방식입니다.

선언형은 무엇을 할 지만 결정하고 구체적인 작업은 컴퓨터가 수행하게 하는 것이죠.
SQL문 같은게 대표적인 예입니다.

보통 패러다임은 그 언어가 추구하는 패러다임을 따라가게 되어 있습니다.
보통은 한가지 패러다임이 아니라 여러 패러다임을 추구하고 그것을 구현할 수 있게 되어 있습니다.
유니티를 한다면 C# 언어를 쓰게 되고 주로 객체지향적, 이벤트 주도적, 함수형, 구조적 방식을 따르는게 좋겠죠.

### 테스트 코드

크게는 동작하는 코드의 로직을 (자동화) 검증을 하는 거라고 생각하시면 좋습니다.
작은 소프트웨어는 한 사람이 어느 정도 시간을 들여 테스트를 하면 어느 정도 검증이 가능하지만 (가능이라고 했지, 좋은 방법은 아닙니다)
여러 사람이 수 개월간 개발한 소프트웨어는 사람이 테스트 하기에는 매우 버겁습니다.
이럴 때는 자동화된 테스트 코드가 반드시 필요하게 됩니다.

또 소프트웨어는 사용자가 쓰면 영원히 쓰게 만들어 줘야 하는데
개발하는 사람은 항상 그 사람이 개발하는 게 아니라 중간에 빠지고 다른 사람이 참여해서 개발하고 하기 때문에
어느 정도 테스트로 검증된 로직에 대한 신뢰성이 확보 되어야 이후 유지보수가 가능해지게 됩니다.

여러 많은 테스트 범위와 기법이 있지만 일반적으로 개발자 레벨에서 많이 생각하고 진행하는 건 단위 테스트입니다.
작은 단위 테스트 코드가 쌓여 소프트웨어를 변경할 때 문제가 없이 변경할 수 있게 해주고, 작동 방식에 대한 문서화도 어느 정도 해주기도 하면서, 기능 단위의 간결한 코드를 짜는데 도움을 줄 수 있습니다.

길게 쓰긴 했는데 테스트 코드를 왜 짜야 하는지 체득할 수 있는 좋은 방법은
좋은 소프트웨어를 만들려고 하는 팀에 속해서 테스트 코드와 함께 기능 개발을 경험해 보는 게 제일 좋습니다.

혹은
이미 개발되어 있는 소프트웨어의 유지보수 인력으로 참여해서
기능 추가를 하거나 버그 수정을 하는데 시간을 보내보면 테스트 코드가 왜 필요한지 알 수도 있습니다.

### 스터디 프로젝트

협업 관련된 branch 전략에 대한 방법을 몰라서 그런 건데

- github에 pull request branch에서 checkout을 통해 새로운 다른 작업을 병행해서 진행하면 별도의 branch로 관리 merge가 가능합니다.
- github에 pull reqeust branch에서 review 중 suggessions 기능을 통해 추가 수정 후 commit 진행이 가능합니다.
